common/repeat_timer.go

```diff
@@ -210,19 +210,23 @@ func (t *RepeatTimer) stop() {
 		panic("Tried to stop a stopped RepeatTimer")
 	}
 	t.ticker.Stop()
+
+	// first we drain send channel
+	drain(t.ticker.Chan())
+	// then we drain receive channel
```

Comment:
is this correct solution for the problem at hand? alternative could be draining the channel on the client side:

```go
timer.Reset() // or timer.Stop()
drain(timer.Chan())
```

Reply:
seems leak test passes without it

Reply:
for me it fails with

```
--- FAIL: TestRepeatTimerReset (5.01s)
        <autogenerated>:1: leaktest: timed out checking goroutines
        <autogenerated>:1: leaktest: leaked goroutine: goroutine 252 [chan send]:
                github.com/tendermint/tmlibs/common.(*RepeatTimer).fireRoutine(0xc4200767d0, 0xc420444f60, 0xc42001b380)
                        /go/src/github.com/tendermint/tmlibs/common/repeat_timer.go:158 +0x94
                created by github.com/tendermint/tmlibs/common.(*RepeatTimer).reset
                        /go/src/github.com/tendermint/tmlibs/common/repeat_timer.go:194 +0x1a9
```

Reply:
The test passes for me if I comment out `drain(t.ticker.Chan())` but fails if I comment `drain(t.ch)`. For you it fails if you comment out either ?

Reply:
hmm

Reply:
> For you it fails if you comment out either ?

yeah

Reply:
Rather than all of this, we just want line 158 to be expanded into yet another select statement that has yet another quit case.

```golang
                select {
                case t_ := <-ch:
-                       t.ch <- t_
+                       select {
+                       case t.ch <- t_:
+                       case <-quit:
+                               return
+                       }
                case <-quit: // NOTE: `t.quit` races.
                        return
                }
```

Reply:
tricky one!

Reply:
another select works even better! awesome, thank you